### DISCUSSION TITLE
FP-like arguments order

### DISCUSSION THREAD

**Discussion by smolak** [gustavoguichard/string-ts#144] [Ideas]
Hi!

Great lib. Just one observation/suggestion: looking at, e.g. `join` function, why have you decided on going with this order of arguments? Asking as you could use 3rd party libs and curry the function with partially applied arguments but it is difficult (impossible?) now, as the argument you apply the thing on comes as first one. E.g. if I wanted to have `snake` utility (joining with `_`), it would be easier to use join like so: `join('_', theDataToApplyThisOn)`.

WDYT?

**Comment by gustavoguichard** [comment]
Hi @smolak , thanks for reaching out!
I'm a big fan of FP and specially currying so I hear you.
The main goal of the library is to replicate the native string fns as much as possible and when building it it felt like `join(thing, '-')` is the closest API for `thing.join('-')`. That was the only reason... do you agree that it resembles the native API a lil more?

That said, I'm open for new ideas and I'm turning the issue into a discussion so we can discuss it and could possibly change that on a future v3 if people agree with it.

Btw, do you know any 3rd party libs that would curry preserving the type inference? I never tried to do so, would love to see the results.

**Reply by gustavoguichard** [reply]
@jly36963 @mjuksel @p9f , opinions?

**Reply by smolak** [reply]
Hi. Got it.
To answer your question - no, I don't know any such lib.
But I checked how Ramda does it, and types for their curry utility functions are defined here: https://github.com/ramda/types/blob/develop/types/curry.d.ts
Perhaps it would be good enough to use it (and I know you don't have currying implemented, at least not yet).
